// toster/lib/tdd.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.suiteTeardown = exports.suiteSetup = exports.teardown = exports.setup = exports.suite = exports.test = undefined;

var _testStatus = require('./test-status');

var _defaultRunner = require('./default-runner');

let test = exports.test = (name, func) => _defaultRunner.runner.addTest(name, func);

let suite = exports.suite = (name, func) => _defaultRunner.runner.addSuite(name, func);

let setup = exports.setup = func => _defaultRunner.runner.addBeforeTestCallback(func);

let teardown = exports.teardown = func => _defaultRunner.runner.addAfterTestCallback(func);

let suiteSetup = exports.suiteSetup = func => _defaultRunner.runner.addBeforeCallback(func);

let suiteTeardown = exports.suiteTeardown = func => _defaultRunner.runner.addAfterCallback(func);

test.skip = (name, func) => _defaultRunner.runner.addTest(name, func, _testStatus.SKIP);

test.only = (name, func) => _defaultRunner.runner.addTestOnly(name, func);

suite.skip = (name, func) => _defaultRunner.runner.addSuite(name, func, _testStatus.SKIP);

suite.only = (name, func) => _defaultRunner.runner.addSuiteOnly(name, func);
