// toster/lib/test-status.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SUITE_FAIL = exports.TestFailStatus = exports.TestTimeoutStatus = exports.SKIP = exports.NOT_EXECUTED = exports.SUCCESS = undefined;

var _util = require('./util');

class TestStatus {
  constructor(name) {
    this.name = name;

    this.duration = -1;
  }

  toString() {
    return this.name;
  }
}

const SUCCESS = exports.SUCCESS = new TestStatus('SUCCESS');
const NOT_EXECUTED = exports.NOT_EXECUTED = new TestStatus('NOT_EXECUTED');
const SKIP = exports.SKIP = new TestStatus('SKIP');

class TestTimeoutStatus extends TestStatus {
  constructor(timeout) {
    super('TIMEOUT');

    this.timeout = timeout;
  }
}

exports.TestTimeoutStatus = TestTimeoutStatus;
class TestFailStatus extends TestStatus {
  constructor(err) {
    super('FAIL');

    this.error = err;
  }

  toJSON() {
    return {
      name: this.name,
      error: (0, _util.serializeError)(this.error)
    };
  }
}

exports.TestFailStatus = TestFailStatus;
const SUITE_FAIL = exports.SUITE_FAIL = new TestFailStatus(new Error('At least one test failed'));
